instantout/reservation/fsm.go

```diff
@@ -0,0 +1,181 @@
+package reservation
+
+import (
+	"context"
+
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightninglabs/loop/fsm"
+	looprpc "github.com/lightninglabs/loop/swapserverrpc"
+)
+
+// Config contains all the services that the reservation FSM needs to operate.
+type Config struct {
+	// Store is the database store for the reservations.
+	Store Store
+
+	// Wallet handles the key derivation for the reservation.
+	Wallet lndclient.WalletKitClient
+
+	// ChainNotifier is used to subscribe to block notifications.
+	ChainNotifier lndclient.ChainNotifierClient
+
+	// ReservationClient is the client used to communicate with the
+	// swap server.
+	ReservationClient looprpc.ReservationServiceClient
+}
+
+// FSM is the state machine that manages the reservation lifecycle.
+type FSM struct {
+	*fsm.StateMachine
+
+	cfg *Config
+
+	reservation *Reservation
+
+	ctx context.Context
+}
+
+// NewFSM creates a new reservation FSM.
+func NewFSM(ctx context.Context, cfg *Config) *FSM {
+	reservationFsm := &FSM{
+		ctx: ctx,
+		cfg: cfg,
+	}
+
+	reservationFsm.StateMachine = fsm.NewStateMachine(
+		reservationFsm.GetReservationStates(), 15,
+	)
+	reservationFsm.ActionEntryFunc = reservationFsm.updateReservation
+
+	return reservationFsm
+}
+
+// NewFSMFromReservation creates a new reservation FSM from an existing
+// reservation recovered from the database.
+func NewFSMFromReservation(ctx context.Context, cfg *Config,
+	reservation *Reservation) *FSM {
+
+	reservationFsm := &FSM{
+		ctx:         ctx,
+		cfg:         cfg,
+		reservation: reservation,
+	}
+
+	reservationFsm.StateMachine = fsm.NewStateMachineWithState(
+		reservationFsm.GetReservationStates(), reservation.State, 15,
+	)
+	reservationFsm.ActionEntryFunc = reservationFsm.updateReservation
+
+	return reservationFsm
+}
+
+// States.
+var (
+	Init                = fsm.StateType("Init")
```

Comment:
nit: I think these states deserve some godocs too. Just some high level explanation what they mean (also the events).